src/foam/nanos/ruler/RulerDAO.js

```diff
@@ -185,37 +175,73 @@ foam.CLASS({
           type: 'Context'
         }
       ],
-      javaCode: `
-        DAO ruleDAO = ((DAO) x.get("ruleDAO")).where(
-          EQ(Rule.DAO_KEY, getDaoKey())
-        );
-        ruleDAO.listen(
-          new UpdateRulesListSink.Builder(x)
-            .setDao(this)
-            .build()
-          , null
-        );
+      javaCode: `DAO ruleDAO = ((DAO) x.get("ruleDAO")).where(
+  EQ(Rule.DAO_KEY, getDaoKey())
+);
+ruleDAO.listen(
+  new UpdateRulesListSink.Builder(getX())
+    .setDao(this)
+    .build()
+  , null
+);
 
-        ruleDAO = ruleDAO.where(
-          EQ(Rule.ENABLED, true)
-        ).orderBy(new Desc(Rule.PRIORITY));
-        addRuleList(ruleDAO, getCreateBefore());
-        addRuleList(ruleDAO, getUpdateBefore());
-        addRuleList(ruleDAO, getRemoveBefore());
-        addRuleList(ruleDAO, getCreateAfter());
-        addRuleList(ruleDAO, getUpdateAfter());
-        addRuleList(ruleDAO, getRemoveAfter());
-      `
+ruleDAO = ruleDAO.where(
+  EQ(Rule.ENABLED, true)
+).orderBy(new Desc(Rule.PRIORITY));
+ruleDAO.select(new AbstractSink(new ReadOnlyDAOContext(getX())) {
+      @Override
+      public void put(Object obj, Detachable sub) {
+        Rule rule = (Rule) obj;
+        rule.setX(getX());
+      }
+    });
+addRuleList(ruleDAO, getCreateBefore());
+addRuleList(ruleDAO, getUpdateBefore());
+addRuleList(ruleDAO, getRemoveBefore());
+addRuleList(ruleDAO, getCreateAfter());
+addRuleList(ruleDAO, getUpdateAfter());
+addRuleList(ruleDAO, getRemoveAfter());`
     },
     {
       name: 'cmd_',
-      javaCode: `
-        if ( PUT_CMD == obj ) {
-          getDelegate().put((FObject) x.get("OBJ"));
-          return true;
-        }
-        return getDelegate().cmd(obj);
-      `
+      javaCode: `if ( PUT_CMD == obj ) {
+  getDelegate().put((FObject) x.get("OBJ"));
+  return true;
+}
+if ( obj instanceof RulerProbe ) {
```

Comment:
Better to do:
if ( ! ( obj instanceof RulerProbe ) ) return getDelegate().cmd_(x, obj);
then have the code below outside of the id.